activation/atxdb.go

```diff
@@ -228,6 +218,43 @@ func (db *ActivationDb) ValidateAtx(atx *types.ActivationTx) error {
 	return nil
 }
 
+// ContextuallyValidateAtx ensures that the previous ATX referenced is the last known ATX for the referenced miner ID.
+// If a previous ATX is not referenced, it validates that indeed there's no previous known ATX for that miner ID.
+func (db *ActivationDb) ContextuallyValidateAtx(atx *types.ActivationTx) error {
+	if atx.PrevATXId != *types.EmptyAtxId {
+		prevAtxIds, err := db.GetNodeAtxIds(atx.NodeId)
+		if err != nil {
+			db.log.WithFields(
+				log.String("atx_id", atx.ShortId()), log.String("node_id", atx.NodeId.Key)).
+				Error("could not fetch node ATXs: %v", err)
+			return fmt.Errorf("could not fetch node ATXs: %v", err)
+		}
+		if len(prevAtxIds) == 0 {
+			return fmt.Errorf("ATX references a previous ATX, but no ATXs were found for this miner ID")
+		}
+		lastAtx := prevAtxIds[len(prevAtxIds)-1]
+		// last atx is not the one referenced
+		if lastAtx != atx.PrevATXId {
+			return fmt.Errorf("last atx is not the one referenced")
+		}
+	} else {
+		prevAtxIds, err := db.GetNodeAtxIds(atx.NodeId)
```

Comment:
The only situation where no prevATX is allowed is if the atx was created in genesis epoch, would you like to add a check for that?

Reply:
Not really. It's always allowed for a miner's first ATX.

Here we check that we don't know of any previous ATXs for this miner, and in the syntactic validation we ensure that the sequence number is 0 in this case. I don't think anything else should be checked.